hazelcast/src/main/java/com/hazelcast/internal/networking/ChannelPipeline.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.networking;
+
+/**
+ *
+ * <h1>Spurious wakeups</h1>
+ * ChannelInboundHandlers/ChannelOutboundHandlers need to be able to deal with
+ * spurious wakeups. E.g. it could be that a PacketDecoder is called with a null
+ * Packet.
+ *
+ * todo: rationale behind this
+ *
+ * <h1>Automatic reprocessing on change</h1>
+ * When a change in the pipeline is detected, the pipeline is automatically
+ * reprocessed. For example when the ProtocolDecoder replaced itself by a
+ * PacketDecoder, the whole pipeline (in this case the PacketDecoder) is
+ * automatically reprocessed.
+ */
+public interface ChannelPipeline {
```

Comment:
Should the pipeline be split in an inbound and outbound pipeline? Internally it is split anyway.